### PR TITLE
Use blobstore for scanner definitions

### PR DESCRIPTION
**Pull request by c-du**
## Description

Move scanner definitions to Blobstore.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why
you did not do so. Valid reasons include, for example, "CI is sufficient",
"No testable changes". Feel free to attach JSON snippets, curl commands,
screenshots.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: central/scannerdefinitions/handler/handler.go [hunk 1]
```diff
@@ -120,7 +120,7 @@ func (h *httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 func (h *httpHandler) get(w http.ResponseWriter, r *http.Request) {
 	// Open the most recent definitions file for the provided `uuid`.
 	uuid := r.URL.Query().Get(`uuid`)
-	f, modTime, err := h.openMostRecentDefinitions(uuid)
+	f, err := h.openMostRecentDefinitions(r.Context(), uuid)
```
Comment (c-du) [inline central/scannerdefinitions/handler/handler.go:123]: I am not sure if this is enough for the authorization. Will look further.

File: central/blob/snapshot/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return
+	}
+	defer func() {
+		if err != nil {
+			_ = os.RemoveAll(tempDir)
+		}
+	}()
+
+	baseName := path.Base(name)
+	// It is definitely not a good practice to have use empty string or all slashes in
```
Comment (dashrews78) [inline central/blob/snapshot/snapshot.go:62]: Should we add those constraints to the blob store now to prevent this from occurring?

Comment (dashrews78) [inline central/blob/snapshot/snapshot.go:62]: Should we add a constraint to the blob store and here to ensure `name` is of a reasonable format?

Comment (c-du) [reply central/blob/snapshot/snapshot.go:62]: We do not have such rule yet. My thinking is to qualify it to be a valid file path, either relative or abs.

Comment (dashrews78) [reply central/blob/snapshot/snapshot.go:62]: I think we should come up with that rule before 4.1 is shipped whether in this PR or another ticket.  IMO.

Comment (c-du) [reply central/blob/snapshot/snapshot.go:62]: Yes. Going to have that in a separate PR. Together I have some other improvement.

File: central/scannerdefinitions/handler/handler_test.go [hunk 3]
```diff
@@ -1,118 +1,162 @@
...
+	datastore datastore.Datastore
+	testDB    *pgtest.TestPostgres
+	tmpDir    string
+}
+
+func TestHandler(t *testing.T) {
+	suite.Run(t, new(handlerTestSuite))
+}
+
+func (s *handlerTestSuite) SetupSuite() {
```
[the remark this answers is not in the document]
Comment (dashrews78) [reply central/scannerdefinitions/handler/handler_test.go:44]: Actually you probably need to add
`//go:build sql_integration`

**Review by dashrews78 — APPROVED** [review]
Fix the one UT issue and it looks fine to me.

**Review by connorgorman — COMMENTED** [review]
Shouldn't these changes be feature flagged? We need a migration right?

**Review by connorgorman — CHANGES_REQUESTED** [review]
I ran Central in offline mode, uploaded scanner definitions, but scanner never ran an update

File: central/scannerdefinitions/handler/handler_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (c-du) [reply central/scannerdefinitions/handler/handler_test.go:44]: Forget to add it for the convenience of goland.

**Comment by c-du** [comment]
> Shouldn't these changes be feature flagged? We need a migration right?

I have a separate PR for migration. I do not want to invest more time on keeping the legacy way working. May be I can merge both PR at the same time.

File: central/scannerdefinitions/handler/handler.go [hunk 4]
```diff
@@ -321,28 +345,35 @@ func (h *httpHandler) openMostRecentDefinitions(uuid string) (file *os.File, mod
...
+
+	defer toClose(onlineFile)
+	offlineFile, err := h.openOfflineBlob(ctx)
 	if err != nil {
-		utils.IgnoreError(onlineFile.Close)
 		return
 	}
+	defer toClose(offlineFile)
 
 	// Return the most recent file, notice that if both don't exist, nil is returned
```
Comment (RTann) [inline central/scannerdefinitions/handler/handler.go:371]: not sure the reason as for why `nil` is returned here is necessarily true anymore

File: central/scannerdefinitions/handler/handler_test.go [hunk 5]
```diff
@@ -125,3 +171,13 @@ func TestServeHTTP_Online_Get(t *testing.T) {
 func mustSetModTime(t *testing.T, path string, modTime time.Time) {
 	require.NoError(t, os.Chtimes(path, time.Now(), modTime))
 }
+
+func (s *handlerTestSuite) mustWriteOffline(content string, modTime time.Time) {
+	modifiedTime, err := types.TimestampProto(modTime)
+	s.NoError(err)
```
Comment (RTann) [inline central/scannerdefinitions/handler/handler_test.go:177]: nit: `s.Require().NoError` here because this means the test writer messed up

**Review by connorgorman — APPROVED** [review]

File: central/scannerdefinitions/handler/handler.go (hunk 1 above)
Comment (c-du) [reply central/scannerdefinitions/handler/handler.go:121]: Augment it when operating blobstore.

File: central/scannerdefinitions/handler/handler.go (hunk 4 above)
Comment (c-du) [reply central/scannerdefinitions/handler/handler.go:371]: It is the same scenario as it was before.

**Pull request merged**